doc/computing/randomness.rst

```diff
@@ -0,0 +1,345 @@
+    is not relevant for our illustration purpose. Also, neither
+    :class:`~sklearn.discriminant_analysis.LinearDiscriminantAnalysis` nor
+    :class:`~sklearn.naive_bayes.GaussianNB` are randomized estimators.
+
+General recommendations
+-----------------------
+
+Getting reproducible results across multiple executions
+.......................................................
+
+In order to obtain reproducible (i.e. constant) results across multiple
+*program executions*, we need to remove all uses of `random_state=None`, which
+is the default. The recommended way is to declare a `rng` variable at the top
+of the program, and pass it down to any object that accepts a `random_state`
+parameter::
+
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn.model_selection import train_test_split
+    >>> import numpy as np
+
+    >>> rng = np.random.RandomState(0)
+    >>> X, y = make_classification(random_state=rng)
+    >>> rf = RandomForestClassifier(random_state=rng)
+    >>> X_train, X_test, y_train, y_test = train_test_split(X, y,
+    ...                                                     random_state=rng)
+    >>> rf.fit(X_train, y_train).score(X_test, y_test)
+    0.84
+
+We are now guaranteed that the result of this script will always be 0.84, no
+matter how many times we run it. Changing the global `rng` variable to a
+different value should affect the results, as expected.
+
+It is also possible to declare the `rng` variable as an integer. This may
+however lead to less significant cross-validation results, as we will see in
+the next section.
+
+.. note::
+    We do not recommend setting the global `numpy` seed by calling
+    `np.random.seed(0)`. See `here
+    <https://stackoverflow.com/questions/5836335/consistently-create-same-random-numpy-array/5837352#comment6712034_5837352>`_
+    for a discussion.
+
+Significance of cross-validation results
+........................................
+
+When we evaluate a randomized estimator performance by cross-validation, we
+want to make sure that the estimator can yield accurate predictions for new
+data, but we also want to make sure that the estimator is robust w.r.t. its
+random initialization. For example, we would like the random weights
+initialization of a :class:`~sklearn.linear_model.SGDCLassifier` to be
+consistently good across all folds: otherwise, when we train that estimator
+on new data, we might get unlucky and the random initialization may lead to
+bad performance. Similarly, we want a random forest to be robust w.r.t the
+set of randomly selected features that each tree will be using.
+
+For these reasons, it is preferable to evaluate the cross-validation
+preformance by letting the estimator use a different RNG on each fold. This
+is done by passing a `RandomState` instance (or None) to the estimator
+initialization.
```

Comment:
I think this reasoning actually breaks in `GridSearchCV`. For a given candidate, I agree that it is preferable to evaluate it with a different RNG on each fold. However, it seems to me that all candidates should be starting from the same RNG. Otherwise the best candidate might just be the one that had lucky seeds.

Basically:

Fold 0 : all candidates use the same RNG
fold 1 : all candidates use the same RNG (different RNG from fold 0)
etc...


`GridSearchCV` does not support that currently.

Thoughts @GaelVaroquaux ?

Reply:
All it tells us is that some improvements in `GridSearchCV` will be due to chance. There's nothing that we can do about this, this is an intrinsic problem of hyper-parameter tuning and of statistics in general, not a breakage of `GridSearchCV`: when we get too close to the optimal, the results are as good as random (for many reasons, including the random choice of initialization and splitting, but also the sampling of the data).

For this reason, there used to be the "1 standard deviation rule", which is to take the simplest model out of all those within one standard deviation of the best (figure 3.7 of the elements of statistical learning, or https://stats.stackexchange.com/questions/80268/empirical-justification-for-the-one-standard-error-rule-when-using-cross-validat). However this rule is really hard to code in a package like scikit-learn.